src/server.c

```diff
@@ -1298,6 +1298,20 @@ uint64_t dictEncObjHash(const void *key) {
     }
 }
 
+/* Return 1 if currently we allow dict to expand. Dict may allocate huge
+ * memory to contain hash buckets when dict expands, that may lead redis
+ * rejects user's requests or evicts some keys, we can stop dict to expand
+ * provisionally if used memory will be over maxmemory after dict expands,
+ * but to gurantee the performance of redis, we still allow dict to expand
+ * if dict load factor exceeds HASHTABLE_MAX_LOAD_FACTOR. */
+int dictExpandAllowed(size_t moreMem, double usedRatio) {
+    if (usedRatio <= HASHTABLE_MAX_LOAD_FACTOR) {
+        return !overMaxmemoryAfterAlloc(moreMem);
```

Comment:
We should probably benchmark the performance impact here when in low memory condition and repeatedly going through `overMaxmemoryAfterAlloc()`, may need some simple caching here if it's big.

Reply:
good idea, but i think it is tricky, i.e. if we allow it to rehash the benchmark will show the impact of rehashing.
maybe to benchmark it we need to create a dummy version with a static response and compare it to the real one.
in the same benchmark maybe it would also be interesting to see the performance implications of an overly populated dict (one that reaches near to the new 1.6 threshold).
@ShooterIT can you please prepare some benchmarks that can measure these concerns (especially the one about repeated calls to overMaxmemoryAfterAlloc?

Reply:
Actually I also ever worried about this question, but i think it doesn't much matter. We need to expand dict only when we add keys. Firstly if redis has child process, we needn't call this function because we disable dict resize. Secondly, let's see the  cost of `overMaxmemoryAfterAlloc`, it will be quick to only call `zmalloc_used_memory` once if memory is sufficient. If memory is not enough, we need to call `freeMemoryGetNotCountedMemory`, we only need to calculate the slaves' output buffer, needn't care about aof rewrite buffer because there is no child process as said above, moreover, `getClientOutputBufferMemoryUsage` is quick, actually we still call many times if we have many replicas, but i think that cost less compared with adding write commands to all replicas' output buffer. To sum up, i don't think `overMaxmemoryAfterAlloc` can bring obvious performance damage. Despite the theory, we still need test data to illustrate, and then decide whether to cache.